### PR TITLE
HARP-12687: User Images are removed from cache when theme is changed

### PR DESCRIPTION
**Pull request by StefanDachwitz**
* remove images from cache only if not used in any other MapView

Signed-off-by: stefan.dachwitz <stefan.dachwitz@here.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

**Review by FraukeF — DISMISSED** [review]
Looks good to me, but as I understand the ticket, but does it solve the issue? Or maybe there is a follow up planned?
Wouldn´t the MapView.userImageCache and the MapView.imageCache register Images referencing the same MapView, and therefore if MapView.imageCache is cleared it would still clear the userImageCache, as described in the ticket?

**Comment by atomicsulfate** [comment]
> Looks good to me, but as I understand the ticket, but does it solve the issue? Or maybe there is a follow up planned?
> Wouldn´t the MapView.userImageCache and the MapView.imageCache register Images referencing the same MapView, and therefore if MapView.imageCache is cleared it would still clear the userImageCache, as described in the ticket?

That's right. This change does not solve the issue.

**Comment by StefanDachwitz** [comment]
That is right, my bad. I caught hung up in fixing the issue when removing shared images (which still happens when the theme is changed).

File: @here/harp-mapview/lib/image/ImageCache.ts [hunk 1]
```diff
@@ -89,28 +89,29 @@ export class ImageCache {
...
+        owner: any,
         url: string,
         imageData?: ImageData | ImageBitmap,
         htmlElement?: HTMLImageElement | HTMLCanvasElement
     ): ImageItem {
         let imageCacheItem = this.findImageCacheItem(url);
         if (imageCacheItem !== undefined) {
-            if (mapView !== undefined && !imageCacheItem.mapViews.includes(mapView)) {
-                imageCacheItem.mapViews.push(mapView);
+            if (owner !== undefined && !imageCacheItem.owners.includes(owner)) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/image/ImageCache.ts:106]: do you need the undefined check? The owner is not optional.

File: @here/harp-mapview/lib/image/ImageCache.ts [hunk 2]
```diff
@@ -120,7 +121,7 @@ export class ImageCache {
                 htmlElement: htmlElement,
                 loaded: false
             },
-            mapViews
+            owners: owners
```
Comment (nzjony) [inline @here/harp-mapview/lib/image/ImageCache.ts:124]:
```suggestion
            owners: [owner]
```

**Review by nzjony — COMMENTED** [review]
LGTM, just some minor comments.

File: @here/harp-mapview/lib/image/ImageCache.ts (hunk 2 above)
Comment (StefanDachwitz) [reply @here/harp-mapview/lib/image/ImageCache.ts:124]: applied your suggestion

File: @here/harp-mapview/lib/image/ImageCache.ts (hunk 1 above)
Comment (StefanDachwitz) [reply @here/harp-mapview/lib/image/ImageCache.ts:106]: removed it

**Comment by StefanDachwitz** [comment]
Reworked the PR to use a proper interface for owners.

**Comment by StefanDachwitz** [comment]
retest this please

**Review by nzjony — APPROVED** [review]

**Pull request merged**